src/Voron/Data/PostingLists/PostingListLeafPage.cs

```diff
@@ -81,29 +82,23 @@ public static void InitLeaf(PostingListLeafPageHeader* header)
         Memory.Copy(newPagePtr, Header, PageHeader.SizeOf);
         InitLeaf(newHeader);
 
-        var existingList = new NativeIntegersList(tx.Allocator, (Header->NumberOfEntries+255)/256 * 256);
-        existingList.Count = ReadAllEntries(tx.Allocator,existingList.RawItems, existingList.Capacity);
+        int numberOfEntriesToReserve = Header->NumberOfEntries + maxAdditionsLimit + maxRemovalsLimit;
+        int requiredSizeAligned256 = (numberOfEntriesToReserve+255)/256 * 256;
+        tempList.EnsureCapacity(requiredSizeAligned256);
 
-        Debug.Assert(existingList.Count == Header->NumberOfEntries);
+        tempList.Count = ReadAllEntries(tx.Allocator,tempList.RawItems, tempList.Capacity);
+        Debug.Assert(tempList.Count == Header->NumberOfEntries);
+        
+        var needSorting = removalsCount > 0 || // any removal force sorting
+                          // here we test if the first new addition is smaller than the largest existing, requiring sorting  
+                          (maxAdditionsLimit > 0 && additions[0] <= tempList.RawItems[tempList.Count - 1]);
 
```

Comment:
This section needs very important comment we want to avoid the "When I wrote this code only god and I knew what it did. Now only god knows".

Reply:
Done